ambry-api/src/main/java/com/github/ambry/quota/RequestCost.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota;
+
+import com.github.ambry.quota.storage.QuotaOperation;
+import java.util.HashMap;
+import java.util.Map;
+
+
+/**
+ * Class representing collection of costs associated with metrics of interest of a request for a resource.
+ */
+public class RequestCost {
+  private final Map<QuotaMetric, Double> costMap;
```

Comment:
@justinlin-linkedin You had this comment on the earlier PR

What I don't get about this class is whether this is a per request or not? Given by it's name and the fields quotaResource and quotaOperation, it seems like every request would have its own RequestCost. Based on that, here is my question:

Why there is a costMap, whose key is QuotaMetric?
What is the difference between QuotaMetric and QuotaResourceType here.

Reply:
This cost is always per request.
When a caller asks for cost of a request, the cost can be of multiple types. Like cost in terms of CAPACITY_UNIT, cost in terms of STORAGE_IN_GB etc. Thats the reason why we have a map of QuotaMetric to cost here. When the QuotaEnforcer returns the cost to the caller, it will specify what is the cost for each metric. The caller can then decide, based on this cost whether to throttle or not. For example, we might say that for delivery cluster its ok to cross the CAPACITY_UNIT quota, but its not ok to cross the STORAGE_IN_GB quota. Also by returning the cost here, we have the option of returning cost of a request to the ambry client in form of response headers.
QuotaResourceType means whether the quota is for container or account or the host level quota. Its is the type of the resource for which quota is enforced. QuotaMetric is the type of quota for the resource. For example, we can have QuotaRequestType as AmbryAccount and QuotaMetric as StorageInGb.